### ISSUE TITLE
PHP7 Collaborator does not implement interface

### ISSUE THREAD

**Issue by sagikazarmark** [phpspec/phpspec#659]
I am getting this error on travis:

`Argument 1 passed to spec\League\Tactician\CommandEvents\Event\CommandExecutedSpec::let() must implement interface League\Tactician\Command, instance of PhpSpec\Wrapper\Collaborator given`

https://travis-ci.org/thephpleague/tactician-command-events/jobs/54361787

**Comment by stof** [comment]
Looks like the hack used by PhpSpec to bypass typehints does not work anymore. Typehint failures are not fatal errors instead of recoverable errors

**Comment by ciaranmcnulty** [comment]
Ugh does this mean we rethink the entire mocking strategy?

**Comment by stof** [comment]
@ciaranmcnulty PhpSpec supports 2 syntaxes to define the type of collaborators:

- through phpdoc (see Prophecy for an example using this standard)
- through hacked typehints thanks to a way found by @everzet to bypass them

The second syntax actually relies on the possibility in PHP to bypass typehints (and so to run broken code) thanks to carefully crafted error handler.

We will probably need to drop this feature on PHP 7

**Comment by ciaranmcnulty** [comment]
I don't think dropping the feature is an option - it's pretty convenient and we can't break existing specs.

I can think of a few hacky avenues to explore, one is to read the specs from disk via a stream wrapper that modifies them to not have type hints.

The conversation should be around which hack is the least painful.

Thanks for reporting this and confirming it's a documented BC issue rather than a bug - we need to have a fix in place by the time 7.0 is released.

This drives me to think any versions released between now and then should have a composer version restriction that disallows 7.0 (we can remove the restriction for CI purposes).

**Comment by sagikazarmark** [comment]
I think most packages which run tests for PHP 7 also allows it to fail. So I would rather leave the tests fail than mess with unresolvable dependencies.

**Comment by mbeccati** [comment]
For the record, I had a chat with Nikita (author of EngineExceptions) and Dmitry (who later changed E_RECOVERABLE_ERRORs to E_ERRORs). Maybe there is still a change to revert the latter and avoid the BC-break here. To be honest the use case here is in fact a way to trick the type hinting system that maybe shouldn't have existed in the first place, but maybe it's worth trying. If you want I can start with an introductory email.

**Comment by docteurklein** [comment]
could it work if one replaces  https://github.com/phpspec/phpspec/blob/7de2100fda6f2b6779cc4afdc399af8b044770af/src/PhpSpec/Runner/Maintainer/ErrorMaintainer.php#L112 with:

``` php
        if (E_ERROR === $level && preg_match($regex, $message, $matches)) {
```

#dumbQuestionMaybe

**Comment by mbeccati** [comment]
@docteurklein E_ERRORs are fatals and can't be intercepted. Exceptions on the other hand could, but then execution won't be resumed in the context PhpSpec is expecting to.

**Comment by ciaranmcnulty** [comment]
Yeah you can't get to the code after `throw` when an exception happens no matter what.

**Comment by sagikazarmark** [comment]
Any idea about this issue's future? If we have to start using dockblocks instead of type hints, it would be better to know sooner than later.

**Comment by sagikazarmark** [comment]
I wonder if the hack which makes it possible to hack type hints could be used for bypassing scalar type hints on PHP < 7.0...

**Comment by ciaranmcnulty** [comment]
There are ways we can make this work with PHP7. Haven't tried hard yet

**Comment by ahilles107** [comment]
Any news on that?

**Comment by nesl247** [comment]
Any update? It's (PHP 7) getting really close to release.

**Comment by digitalkaoz** [comment]
its working with phpdoc as @stof stated...but its not the sweetest solution

**Comment by digitalkaoz** [comment]
i wrote a small tool to convert between typehints and docblocks...using it on travis only under php7. works great but is still in progress

**Comment by stof** [comment]
@digitalkaoz there is a work in progress by @ciaranmcnulty to support the typehint-based notation on PHP 7 too.

**Comment by digitalkaoz** [comment]
@stof that would be even more great, until then https://github.com/digitalkaoz/typehint-to-docblock can jump in...planning a phar which is easily downloadable on travis

**Comment by ryanwinchester** [comment]
@digitalkaoz thanks, that is working for me at the moment.

**Comment by ciaranmcnulty** [comment]
Work in progress is here: #788

**Comment by sagikazarmark** [comment]
Is this solved now? When can we expect a stable release?

**Comment by digitalkaoz** [comment]
its merged, so if you install master it should work

**Comment by ciaranmcnulty** [comment]
The fix is present in 2.4.0-alpha and we are actively asking users to report any remaining issues.

If all goes well we will go through beta,etc,stable releases over the next few weeks

**Comment by sagikazarmark** [comment]
Awesome, thanks. Closing it then.
